pkg/admission/customresourcevalidation/oauth/validate_github.go

```diff
@@ -0,0 +1,63 @@
+package oauth
+
+import (
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	configv1 "github.com/openshift/api/config/v1"
+	crvalidation "github.com/openshift/origin/pkg/admission/customresourcevalidation"
+)
+
+func ValidateGitHubIdentityProvider(provider *configv1.GitHubIdentityProvider, challenge bool, mappingMethod configv1.MappingMethodType, fieldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+	if provider == nil {
+		errs = append(errs, field.Required(fieldPath, ""))
+		return errs
+	}
+
+	errs = append(errs, ValidateOAuthIdentityProvider(provider.ClientID, provider.ClientSecret, fieldPath.Child("provider"))...)
+
+	if challenge {
+		errs = append(errs, field.Invalid(fieldPath.Child("challenge"), challenge, "A GitHub identity provider cannot be used for challenges"))
+	}
+
+	if len(provider.Teams) > 0 && len(provider.Organizations) > 0 {
+		errs = append(errs, field.Invalid(fieldPath.Child("organizations"), provider.Organizations, "specify organizations or teams, not both"))
+		errs = append(errs, field.Invalid(fieldPath.Child("teams"), provider.Teams, "specify organizations or teams, not both"))
```

Comment:
I see that this is coming from the old validation logic, but it doesn't look like there's any reason you _couldn't_ have both orgs and teams. I could see this being useful, like `"organizations": ["keycloak"], "teams": ["openshift/team-red-hat"]`. Maybe that's confusing though?

Reply:
The current oauth-server code would not handle such situations and would fail right away after it noticed that the user is not a part of the `keycloak` organization:

https://github.com/openshift/origin/blob/9c28792c0aece4f9b7926746949111faf61a3be7/pkg/oauthserver/oauth/external/github/github.go#L181-L202

Not sure what lead to such design decision, I bet there is a good reason though, maybe @enj would know.

Reply:
The user could be a part of the `keycloak` organization. For instance, this would be valid for me: `"organizations": ["quay"], "teams": ["coreos//coreos-full-time"]`

Reply:
This has never been requested as a "feature" nor do I consider it valuable.  It does not warrant a change to osin and we can always relax the validation later if we ever need it.